xds/internal/xdsclient/e2e_test/lds_watchers_test.go

```diff
@@ -0,0 +1,797 @@
+			notWatchedResource:     e2e.DefaultClientListener("unsubscribed-lds-resource", rdsNameNewStyle),
+			wantUpdate: xdsresource.ListenerUpdateErrTuple{
+				Update: xdsresource.ListenerUpdate{
+					RouteConfigName: rdsNameNewStyle,
+					HTTPFilters:     []xdsresource.HTTPFilter{{Name: "router"}},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			// Register a watch for a listener resource and have the watch
+			// callback push the received update on to a channel.
+			updateCh := testutils.NewChannel()
+			ldsCancel := client.WatchListener(test.resourceName, func(u xdsresource.ListenerUpdate, err error) {
+				updateCh.Send(xdsresource.ListenerUpdateErrTuple{Update: u, Err: err})
+			})
+
+			// Configure the management server to return a single listener
+			// resource, corresponding to the one we registered a watch for.
+			resources := e2e.UpdateOptions{
+				NodeID:         nodeID,
+				Listeners:      []*v3listenerpb.Listener{test.watchedResource},
+				SkipValidation: true,
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			if err := mgmtServer.Update(ctx, resources); err != nil {
+				t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+			}
+
+			// Verify the contents of the received update.
+			if err := verifyListenerUpdate(ctx, updateCh, test.wantUpdate); err != nil {
+				t.Fatal(err)
+			}
+
+			// Configure the management server to return an additional listener
+			// resource, one that we are not interested in.
+			resources = e2e.UpdateOptions{
+				NodeID:         nodeID,
+				Listeners:      []*v3listenerpb.Listener{test.watchedResource, test.notWatchedResource},
+				SkipValidation: true,
+			}
+			if err := mgmtServer.Update(ctx, resources); err != nil {
+				t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+			}
+			if err := verifyNoListenerUpdate(ctx, updateCh); err != nil {
+				t.Fatal(err)
+			}
+
+			// Cancel the watch and update the resource corresponding to the original
+			// watch.  Ensure that the cancelled watch callback is not invoked.
+			ldsCancel()
+			resources = e2e.UpdateOptions{
+				NodeID:         nodeID,
+				Listeners:      []*v3listenerpb.Listener{test.updatedWatchedResource, test.notWatchedResource},
+				SkipValidation: true,
+			}
+			if err := mgmtServer.Update(ctx, resources); err != nil {
+				t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+			}
+			if err := verifyNoListenerUpdate(ctx, updateCh); err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
+// TestLDSWatch_TwoWatchesForSameResourceName covers the case where two watchers
+// exist for a single listener resource.  The test verifies the following
+// scenarios:
+// 1. An update from the management server containing the resource being
+//    watched, results in the invocation of both watch callbacks
```

Comment:
Similar to above, s/results to should result and s/does not result/should not result and add periods.